File: code/sorting/pigeonhole_sort/PigeonHoleSort.cs
// Part of OpenGenus Cosmos
using System.IO;
using System;
using System.Linq;
using System.Collections.Generic;

class PigeonHoleSort
{

    static void InsertValuesInPigeonHoles(int[] unsortedArray, List<int>[] pigeonHoles){
        var minValue = unsortedArray.Min();
        foreach(var item in unsortedArray){
            var i = item - minValue;
            if(pigeonHoles[i] == null){
                pigeonHoles[i] = new List<int>();
            }
            pigeonHoles[i].Add(item);
        }
    }

    static int[] ExtractSortedValuesFromPigeonHoles(List<int>[] pigeonHoles){
        var sortedList = new List<int>();
        foreach (var hole in pigeonHoles)
        {
            foreach (var holeItem in hole)
            {
                sortedList.Add(holeItem);
            }
        }
        return sortedList.ToArray();
    }

    static int[] sortUsingPigeonHole(int[] unsortedArray){
        var rangeOfNumbers = unsortedArray.Max() - unsortedArray.Min() + 1;
        var pigeonHoles = new List<int>[rangeOfNumbers];
        InsertValuesInPigeonHoles(unsortedArray, pigeonHoles);
        return ExtractSortedValuesFromPigeonHoles(pigeonHoles);
    }

    static void DisplaySortedItems(int[] sortedArray){
        Console.WriteLine("Sorted array");
        foreach (var item in sortedArray)
        {
            Console.Write($"{item} ");
        }
    }

    static int[] ConvertToIntArray(string[] arr){
        var length = arr.Length;
        var intArray = new int[length];
        for(int i=0;i<length;i++){
            intArray[i] = int.Parse(arr[i]);
        }
        return intArray;
    }

    static void Main()
    {
        Console.WriteLine("Enter the unsorted elements separated by space");
        var unsortedArray = Console.ReadLine().Split(' ');
        var sortedArray = sortUsingPigeonHole(ConvertToIntArray(unsortedArray));
        DisplaySortedItems(sortedArray);
    }
}

File: code/sorting/src/pigeonhole_sort/pigeonhole_sort.php
<?php

/* Part of Cosmos by OpenGenus Foundation */

/* Example Usage:
	print_r(pigeonholeSort([9, 2, 17, 5, 16]));
*/

function pigeonholeSort($array) {
	$min = min($array);
	$max = max($array);
	$numHoles = $max - $min + 1;

	$holes = array_fill(0, $numHoles, 0);

	foreach ($array as $item) {
		$holes[$item - $min] += 1;
	}

	$sortedArray = [];

	for ($i = 0; $i < count($holes); $i++) {
		while ($holes[$i] > 0) {
			array_push($sortedArray, $i + $min);
			$holes[$i] -= 1;
		}
	}

	return $sortedArray;

}
